### PR TITLE
Enable `torch.save()` + `torch.load()` for different `map_location`s

### LINKED ISSUE #158: Serialization: `torch.load(..., map_location="cpu")` doesn't map all tensors in `Laplace` instance
Hi

I came across an issue when serializing a `BaseLaplace`-derived object with tensors mapped to GPU and then reading it back in and mapping to CPU only. The same goes for a data structure containing such an object. See below for details on the use case.

When using `torch.load(..., map_location="cpu")` in

```py
import dill

lap = Laplace(
    model=model.to("cuda"),
    likelihood="regression",
    ...
)
lap.fit(train_dl)
torch.save(lap, "lap_gpu.pt", pickle_module=dill)

lap = torch.load("lap_gpu.pt", map_location="cpu")
```

one assumes that all tensors are mapped to CPU. This is true for weight tensors in `lap.model`, for instance. However, some tensor-valued properties in `Laplace`, such as `prior_precision_diag`, are calculated each time when accessed, and some of them access `self._device` for that. The latter is defined once in the `BaseLaplace` constructor

```py
self._device = next(model.parameters()).device
```

to be the same device as the one `model` lives on. `torch.load(..., map_location="cpu")` won't affect `self._device` and thus for all properties which use `self._device`, an attempt to calculate them will mix GPU and CPU tensors, which will fail.

One solution is to make `_device` a property.

```py
@property
def _device(self):
    return next(self.model.parameters()).device
```

I have a branch with this working. Before I open a PR, I'd like to know whether you think this is useful and if there are other implications that I did not encounter so far (test suite passes, of course). At least in terms of performance, this should be ok:

```py
>>> %timeit next(lap.model.parameters()).device
1.87 µs ± 2.73 ns per loop (mean ± std. dev. of 7 runs, 1,000,000 loops each)
```

Here are the combinations of `hessian_structure` and `subset_of_weights` that are affected (`load_ok=no`) or not (`load_ok=yes`).

```
hessian_structure subset_of_weights                                                             err load_ok
             diag        last_layer ValueError: Jacobians need to be on the same device as Laplace.      no
             kron        last_layer                                                                     yes
             full        last_layer                                                                     yes
             diag               all ValueError: Jacobians need to be on the same device as Laplace.      no
          lowrank               all                         RuntimeError: No HIP GPUs are available      no
             kron               all                                                                     yes
             full               all                         RuntimeError: No HIP GPUs are available      no
             full        subnetwork                         RuntimeError: No HIP GPUs are available      no
```

With the proposed fix, all of these work. I can provide more details and code to reproduce the above if needed.

The use case is serializing a fitted `Laplace` instance, including `self.model`, to disk and avoid the usual torch `load_state_dict` gymnastics (I'm aware of #148, but for small enough models and many test calculations, serializing stuff to one blob and loading it back in is more convenient). I may then want to run prediction later on other machines (in a cluster) by loading a fitted `Laplace` from disk and where CPU execution is fast enough. Disclaimer: I'm aware that serializing complex objects is not the best of practices because it has lots of potential pitfalls, and that this is not recommended for long-term data storage.

Still, I think it would be nice if `Laplace` instances behaved transparently when using `torch.load()` as `torch.Module`s do.

Thanks.

### PR DESCRIPTION
**Pull request by elcorto**
Closes #158

### CODE REVIEW

File: tests/test_serialization.py
Comment (elcorto) [inline tests/test_serialization.py:304]: We ignore the `self.X.device` check  on `LLLaplace` instances because of the reason described in the comment. Feedback from a core dev should clarify if this is OK or how else to handle this.

File: tests/test_serialization.py
Comment (elcorto) [inline tests/test_serialization.py:257]: The usage of `dill` for LLLaplace to pass tests is optional and those will be skipped with a message. Is this OK? I'd not make `dill` a requirement, but maybe the developers would like to see a warning here in addition to the skip?

**Review by elcorto — COMMENTED** [review]
The PR is marked as WIP because of two questions I have in `test_serialization.py::test_map_location`. Thanks.

File: tests/test_serialization.py
Comment (wiseodd) [inline tests/test_serialization.py:273]: Does this line create an actual file? If so, it should be cleaned up after this unit test finishes. You can add this file to the `cleanup` function above in this file.

File: tests/test_serialization.py
Comment (wiseodd) [reply tests/test_serialization.py:304]: `self.X` is only being saved to infer the last-layer of the `self.model`. If I understand correctly, the `self.model` here is also serialized, right? In this case, we don't need to care about `self.X`, so I'm fine with skipping.

**Review by wiseodd — CHANGES_REQUESTED** [review]
Thanks for the PR! I left some comments. Happy to merge after they're addressed.

File: tests/test_serialization.py
Comment (elcorto) [reply tests/test_serialization.py:304]: Yes, `self.model` is serialized -- this is why `next(self.model.parameters()).device` returns the correct device. In the code, I only see `self.X` used [here](https://github.com/aleximmer/Laplace/blob/2543274aad34e2612d708a58c55272fadbc41626/laplace/lllaplace.py#L109-L113) and [here](https://github.com/aleximmer/Laplace/blob/2543274aad34e2612d708a58c55272fadbc41626/laplace/lllaplace.py#L181-L185) where it is sent to `self._device` if needed.

File: tests/test_serialization.py
Comment (elcorto) [reply tests/test_serialization.py:257]: The supported API so far is `torch`-compatible as in

```py
torch.save(obj, filename)
```

So, `obj`, being a `Laplace` instance, would need to

* figure out that it is being serialized
* in that case, catch any `AttributeError` such as

  ```
  AttributeError: Can't pickle local object 'FeatureExtractor._get_hook.<locals>.hook'
  ```

  Here one must make the assumption that any `AttributeError` is actually caused by an attempted `torch.save()` call.

I understand the reasoning behind your suggestion. However I don't know, off the top of my head, how to implement this. I'm happy for any hints here.

One solution that comes to mind would be to have a `laplace.save()` function that wraps `torch.save()` but extends the API, kind of breaks `torch` compatibility and needs to be documented.

The other option is to document potential serialization fails with plain `pickle` and advise users to use `torch.save(..., pickle_module=dill)`. In either case, users must be made aware via documentation.

File: tests/test_serialization.py
Comment (elcorto) [reply tests/test_serialization.py:230]: Sure no problem. I did not find a coding style enforced (e.g. by `pre-commit`) or documented (I may have overlooked that), so I `black`-formatted my changes (`ruff format` actually). Please let me know where the coding rules are documented such that I can make changes accordingly.

File: tests/test_serialization.py
Comment (elcorto) [reply tests/test_serialization.py:273]: It uses the `pytest` [`tmp_path` fixture])(https://docs.pytest.org/en/latest/reference/reference.html#tmp-path) so the file is subject to `pytest`'s file retention policy.

File: tests/test_serialization.py
Comment (wiseodd) [reply tests/test_serialization.py:304]: Then let's go forward with ignoring `self.X.device` as you've already done.

File: tests/test_serialization.py
Comment (wiseodd) [reply tests/test_serialization.py:257]: I'm leaning towards the documentation-based approach.

How about writing a new example specifically for serialization on README.md? Currently the Laplace-specific serialization example is coupled with another example:(https://github.com/aleximmer/Laplace/blob/2543274aad34e2612d708a58c55272fadbc41626/README.md?plain=1#L43-L76

Let's extract the serialization code from there, and add your `torch`-native-but-not-portable alternative. We can put this new example after  the `### Applying the LA over only a subset of the model parameters` example.

File: tests/test_serialization.py
Comment (wiseodd) [reply tests/test_serialization.py:230]: We don't have a strict coding standard other than a single quote policy. IIRC we discussed this internally during the conception. 

Feel free to run formatting, but please run `black` with `--skip-string-normalization` when you do formatting.

File: tests/test_serialization.py
Comment (wiseodd) [reply tests/test_serialization.py:273]: Ok good to know!

File: tests/test_serialization.py
Comment (elcorto) [reply tests/test_serialization.py:304]: Agreed.

File: tests/test_serialization.py
Comment (elcorto) [reply tests/test_serialization.py:230]: OK, will do, thanks.

File: tests/test_serialization.py
Comment (elcorto) [reply tests/test_serialization.py]: I added a section to the README in 807644071078e8fb3bf41312c695dd1e762f4fa9 as you suggested. I feel that this may be a bit too technical for the README, should this instead go into the docs over at https://aleximmer.github.io/Laplace/?

**Review by wiseodd — APPROVED** [review]

**Comment by wiseodd** [comment]
Thanks, @elcorto! Everything looks good. I'm fine with putting the explanation on the README---currently our docs website is admittedly not very readable.

I'm merging this --- thanks again for your contrib.

**Pull request merged**
